cluster/src/main/java/org/apache/iotdb/cluster/query/ClusterPlanExecutor.java

```diff
@@ -81,10 +97,48 @@ public QueryDataSet processQuery(PhysicalPlan queryPlan, QueryContext context)
   }
 
   @Override
-  protected List<String> getNodesList(String schemaPattern, int level) {
-    // TODO-Cluster: enable meta queries
-    throw new UnsupportedOperationException("Not implemented");
-    //return metaGroupMember.getNodeList(schemaPattern, level);
+  protected List<String> getNodesList(String schemaPattern, int level)
+      throws IOException, TException, InterruptedException, MetadataException {
+    Set<String> nodeListSet = new HashSet<>(
+        MManager.getInstance().getNodesList(schemaPattern, level));
+
+    GetNodesListHandler handler = new GetNodesListHandler();
+
+    ExecutorService pool = new ScheduledThreadPoolExecutor(THREAD_POOL_SIZE);
+
+    for (Node node : metaGroupMember.getAllNodes()) {
+      if (node.equals(metaGroupMember.getThisNode())) {
+        continue;
+      }
+      pool.submit(() -> {
+        DataClient client = null;
+        try {
+          client = metaGroupMember.getDataClient(node);
+        } catch (IOException e) {
+          logger.error("Failed to connect to node: {}", node, e);
+        }
+        AtomicReference<List<String>> response = new AtomicReference<>(null);
+        handler.setResponse(response);
+        handler.setContact(node);
+        synchronized (response) {
+          try {
+            if(client!= null){
+              client.getNodeList(null, schemaPattern, level, handler);
+              response.wait(connectionTimeoutInMS);
+            }
+          } catch (TException e) {
+            logger.error("Error occurs when getting node lists in node {}.", node, e);
+          } catch (InterruptedException e) {
+            logger.error("Interrupted when getting node lists in node {}.", node, e);
+          }
+        }
+        List<String> paths = response.get();
+        if (paths != null) {
+          nodeListSet.addAll(response.get());
```

Comment:
As `nodeListSet` is modified in different threads, maybe you should use a concurrent class.

By the way, the naming `nodeListSet` is confusing, probably better to just use `nodeSet`.

Reply:
Thanks for your reminding. `ConcurrentSkipListSet` is used instead of `Set` now.